File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        res = []
        def inorder_dfs(root):
            if not root:
                return
            inorder_dfs(root.left)
            res.append(root.val)
            inorder_dfs(root.right)
        inorder_dfs(root)
        for i in range(1, len(res)):
            if res[i] <= res[i-1]:
                return False
        return True

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int goodNodesCount;
    public int goodNodes(TreeNode root) {
        this.goodNodesCount = 0;
        if(root == null) return goodNodesCount;
        calculateGoodNodes(root, Integer.MIN_VALUE);
        return goodNodesCount;
    }
    private void calculateGoodNodes(TreeNode node, int maxVal){
        if(node == null) return;
        if(node.val >= maxVal){
            goodNodesCount++;
        }
        int newMaxVal = Math.max(maxVal, node.val);
        calculateGoodNodes(node.left, newMaxVal);
        calculateGoodNodes(node.right, newMaxVal);
    }
}
